icu4c/source/common/ucnv.cpp

```diff
@@ -1803,10 +1811,18 @@ ucnv_toUChars(UConverter *cnv,
     }
     if(srcLength>0) {
         srcLimit=src+srcLength;
+
+        // Trim destCapacity such that computing the destLimit cannot
+        // overflow and wrap the pointer around. Which would be undefined behavior,
+        // meaning that compilers can optimize away attempts to check for it.
+        int32_t pinnedDestCapacity = (UChar *)U_MAX_PTR(dest) - (UChar *)dest;
+        if (pinnedDestCapacity < destCapacity) {
+            destCapacity = pinnedDestCapacity;
+        }
         destLimit=dest+destCapacity;
 
-        /* pin the destination limit to U_MAX_PTR; NULL check is for OS/400 */
-        if(destLimit<dest || (destLimit==NULL && dest!=NULL)) {
```

Comment:
if (desLimit<dest) was the source of the problems. The compiler optimizes it away, because destCapacity is known >= 0, meaning, in the absence of overflow, the condition must be true. Overflow is undefined behavior, meaning anything goes.